### PR TITLE
LOK-2188: Add first observation time to calculate reachability

### PR DESCRIPTION
**Pull request by cgorantla**
Consider first observation time while calculating reachability

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2188

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitoredStatusService.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+public class MonitoredStatusService {
+
+   private final MonitoredServiceStateRepository monitoredServiceStateRepository;
+   private final MonitoredServiceStatusMapper monitoredServiceStatusMapper;
+
+    public Optional<MonitoredServiceStatusDTO> getServiceStatus(String tenantId, long serviceId) {
```
Comment (bouff) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitoredStatusService.java:46]: Nit: formatting (leading space)

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/GrpcConfig.java [hunk 2]
```diff
@@ -86,9 +86,20 @@ public GrpcServerManager startServer(MonitoringLocationGrpcService locationGrpc,
                                          InventoryServerInterceptor interceptor,
                                          ActiveDiscoveryGrpcService activeDiscoveryGrpcService,
                                          IcmpActiveDiscoveryGrpcService icmpActiveDiscoveryGrpcService,
-                                         PassiveDiscoveryGrpcService passiveDiscoveryGrpcService) {
+                                         PassiveDiscoveryGrpcService passiveDiscoveryGrpcService,
+                                         MonitorServiceStatusGrpcService monitorServiceStatusGrpcService,
```
Comment (bouff) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/GrpcConfig.java:90]: Small nit: for consistency - should this be named `MonitoredServiceStatus...` (given `MonitoredService...`)

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcLocationService.java [hunk 3]
```diff
@@ -99,5 +99,5 @@ public Mono<Boolean> deleteLocation(@GraphQLArgument(name = "id") long id, @Grap
         certificateManagerClient.revokeCertificate(tenantId, id, accessToken);
         return Mono.just(status);
     }
-
+
```
Comment (bouff) [inline rest-server/src/main/java/org/opennms/horizon/server/service/GrpcLocationService.java:102]: Nit: remove space here

File: rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredServiceStatus.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+@Getter
+@Setter
+public class MonitoredServiceStatus {
+
+    private long id;
+    private String tenant_id;
+    private long monitored_service_id;
+    private boolean state;
+    private long first_observation_time;
```
Comment (bouff) [inline rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredServiceStatus.java:42]: Nit: code style: all other model entities are camel case

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/TSDBMetricsService.java [hunk 5]
```diff
@@ -109,6 +109,18 @@ public Mono<TimeSeriesQueryResult> getMetric(@GraphQLEnvironment ResolutionEnvir
...
+                                                 String name, Map<String, String> labels,
+                                                 Integer timeRange, TimeRangeUnit timeRangeUnit, Map<String,String> optionalParams) {
+
+        Map<String, String> metricLabels = Optional.ofNullable(labels)
+            .map(HashMap::new).orElseGet(HashMap::new);
+
+        String tenantId = headerUtil.extractTenant(env);
+        String queryString = queryService.getCustomQueryString(name, metricLabels, timeRange, timeRangeUnit, optionalParams);
+        return getMetrics(tenantId, queryString);
+
```
Comment (bouff) [inline rest-server/src/main/java/org/opennms/horizon/server/service/metrics/TSDBMetricsService.java:122]: Nit: remove trailing newline

File: shared-lib/inventory/src/main/proto/monitoredService.proto [hunk 6]
```diff
@@ -39,3 +39,14 @@ message MonitoredServiceDTO {
   int64 monitored_service_type_id = 3;
   int64 ip_interface_id = 4;
 }
+
+message MonitoredServiceQuery {
+  string  tenant_id = 1;
```
Comment (bouff) [inline shared-lib/inventory/src/main/proto/monitoredService.proto:44]: Nit: extra space after `string` here and next line

File: rest-server/src/main/java/org/opennms/horizon/server/service/MonitorStatusGrpcService.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    private final ServerHeaderUtil headerUtil;
+    private final MonitoredServiceStatusMapper mapper;
+    private final InventoryClient client;
+
+    @GraphQLQuery
+    public Mono<MonitoredServiceStatus> getMonitorStatus(@GraphQLArgument(name = "request") MonitoredServiceStatusRequest request,
+                                                         @GraphQLEnvironment ResolutionEnvironment env) {
+       var monitorStatusProto = client.getMonitorStatus(request, headerUtil.getAuthHeader(env));
+       var monitoredServiceStatus = mapper.protoToModel(monitorStatusProto);
+       return  Mono.just(monitoredServiceStatus);
```
Comment (bouff) [inline rest-server/src/main/java/org/opennms/horizon/server/service/MonitorStatusGrpcService.java:57]: Nit: extra space after `return`

**Review by bouff — COMMENTED** [review]
Just a few minor comments.

File: rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredServiceStatus.java (hunk 4 above)
Comment (cgorantla) [reply rest-server/src/main/java/org/opennms/horizon/server/model/inventory/MonitoredServiceStatus.java:42]: oops, I copied from the proto and forgot to update to camelCase.

**Review by bouff — APPROVED** [review]

**Review by freddy61025 — APPROVED** [review]
LGTM

**Pull request merged**